Klib validation failing on CI but not locally

```
* What went wrong:	
Execution failed for task ':redwood-layout-modifiers:klibApiMerge'.	
> java.io.FileNotFoundException: File does not exist: /Users/runner/work/redwood/redwood/redwood-layout-modifiers/build/api/js/redwood-layout-modifiers.klib.api	
```
- Project: https://github.com/cashapp/redwood
- PR: https://github.com/cashapp/redwood/pull/1884
- SHA: 5dc8c3377b3c3b7760416dcc5308033090213ca3

Attempt 1
 - Action log: https://github.com/cashapp/redwood/actions/runs/8360157932/job/22885183135?pr=1884
 - Build scan: https://gradle.com/s/p5ha2lzccuwiq

Attempt 2
 - Action log: https://github.com/cashapp/redwood/actions/runs/8360157932/job/22891922690?pr=1884
 - Build scan: https://gradle.com/s/ehehhmdxrrf5a

Attempt 3
 - Action log: https://github.com/cashapp/redwood/actions/runs/8360157932/job/22893510501?pr=1884
 - Build scan: https://gradle.com/s/tiijq75sgyt4c

I have not been able to reproduce this locally on my Mac. I tried executing the same command (after a `clean`) with `--max-workers=3` to match the core count of the CI machine but it always succeeds (5 attempts).

I also don't know why it's always this specific module and the JS target that fails. I apply it to ~40 modules in this build, most of which are multiplatform containing JS targets. Very weird.

It's possible the problem can be identified just from looking at the code of the plugin, but I don't have time to do that right now. I can try investigating more later tonight or tomorrow, but figured I'd file the issue first.

I get a similar exception on Windows for [this project](https://github.com/kordlib/kord/tree/76c5474aef9eaee80df752084799e80882d5363d) when klib validation is enabled [here](https://github.com/kordlib/kord/blob/76c5474aef9eaee80df752084799e80882d5363d/buildSrc/src/main/kotlin/BinaryCompatibility.kt) (by adding `klib.enabled = true`):
```
> Task :core:live-tests:klibApiMergeInferred FAILED                                                                                                                                                                                                                                                                           

FAILURE: Build failed with an exception.                                                                                                                                                                                                                                                                                      

* What went wrong:
Execution failed for task ':core:live-tests:klibApiMergeInferred'.
> java.io.FileNotFoundException: File does not exist: C:\...\kord\core\live-tests\build\api\js\live-tests.klib.api 
```
The relevant parts of the project layout look like this:
```
kord (root project, BCV plugin not applied)
|
+-- core (public api subproject, BCV plugin applied)
    |
    +-- live-tests (internal subproject, BCV plugin not applied)
```
It seems like BCV tries to find the `.klib.api` file for a subproject that won't ever create one because the BCV plugin isn't applied there.

The workaround I found is putting this into `kord/core/build.gradle.kts`:
```kotlin
apiValidation {
    ignoredProjects += "live-tests"
}
```

--- a/src/functionalTest/kotlin/kotlinx/validation/api/Assert.kt
+++ b/src/functionalTest/kotlin/kotlinx/validation/api/Assert.kt
@@ -24,6 +24,13 @@ internal fun BuildResult.assertTaskFailure(task: String) {
     assertTaskOutcome(TaskOutcome.FAILED, task)
 }
 
+/**
+ * Helper `fun` for asserting a [TaskOutcome] to be equal to [TaskOutcome.SKIPPED]
+ */
+internal fun BuildResult.assertTaskSkipped(task: String) {
+    assertTaskOutcome(TaskOutcome.SKIPPED, task)
+}
+
 private fun BuildResult.assertTaskOutcome(taskOutcome: TaskOutcome, taskName: String) {
     assertEquals(taskOutcome, task(taskName)?.outcome)
 }

--- a/src/functionalTest/kotlin/kotlinx/validation/test/KlibVerificationTests.kt
+++ b/src/functionalTest/kotlin/kotlinx/validation/test/KlibVerificationTests.kt
@@ -18,6 +18,7 @@ import org.junit.Test
 import java.io.File
 import java.nio.file.Files
 import java.nio.file.Paths
+import kotlin.test.assertFalse
 import kotlin.test.assertTrue
 
 internal const val BANNED_TARGETS_PROPERTY_NAME = "binary.compatibility.validator.klib.targets.disabled.for.testing"
@@ -632,5 +633,40 @@ internal class KlibVerificationTests : BaseKotlinGradleTest() {
                         "androidNativeX86, linuxArm64.linux, linuxX64, mingwX64]"
             )
         }
+    }
+
+    @Test
+    fun `apiDump should not fail for empty project`() {
+        val runner = test {
+            baseProjectSetting()
+            addToSrcSet("/examples/classes/AnotherBuildConfig.kt", sourceSet = "commonTest")
+            runApiDump()
+        }
+
+        runner.build().apply {
+            assertTaskSkipped(":klibApiDump")
+        }
+        assertFalse(runner.projectDir.resolve("api").exists())
+    }
+
+    @Test
+    fun `apiDump should not fail if there is only one target`() {
+        val runner = test {
+            baseProjectSetting()
+            addToSrcSet("/examples/classes/AnotherBuildConfig.kt", sourceSet = "commonTest")
+            addToSrcSet("/examples/classes/AnotherBuildConfig.kt", sourceSet = "linuxX64Main")
+            runApiDump()
+        }
+        checkKlibDump(runner.build(), "/examples/classes/AnotherBuildConfig.klib.linuxX64Only.dump")
+    }
+
+    @Test
+    fun `apiCheck should not fail for empty project`() {
+        val runner = test {
+            baseProjectSetting()
+            addToSrcSet("/examples/classes/AnotherBuildConfig.kt", sourceSet = "commonTest")
+            runApiCheck()
+        }
+        runner.build()
     }
 }

--- a/src/functionalTest/resources/examples/classes/AnotherBuildConfig.klib.linuxX64Only.dump
+++ b/src/functionalTest/resources/examples/classes/AnotherBuildConfig.klib.linuxX64Only.dump
@@ -1,0 +1,14 @@
+// Klib ABI Dump
+// Targets: [linuxX64]
+// Rendering settings:
+// - Signature version: 2
+// - Show manifest properties: true
+// - Show declarations: true
+
+// Library unique name: <testproject>
+final class org.different.pack/BuildConfig { // org.different.pack/BuildConfig|null[0]
+    constructor <init>() // org.different.pack/BuildConfig.<init>|<init>(){}[0]
+    final fun f1(): kotlin/Int // org.different.pack/BuildConfig.f1|f1(){}[0]
+    final val p1 // org.different.pack/BuildConfig.p1|{}p1[0]
+        final fun <get-p1>(): kotlin/Int // org.different.pack/BuildConfig.p1.<get-p1>|<get-p1>(){}[0]
+}

--- a/src/main/kotlin/BinaryCompatibilityValidatorPlugin.kt
+++ b/src/main/kotlin/BinaryCompatibilityValidatorPlugin.kt
@@ -10,8 +10,8 @@ import kotlinx.validation.api.klib.konanTargetNameMapping
 import org.gradle.api.*
 import org.gradle.api.plugins.*
 import org.gradle.api.provider.*
+import org.gradle.api.specs.Spec
 import org.gradle.api.tasks.*
-import org.jetbrains.kotlin.cli.jvm.main
 import org.jetbrains.kotlin.gradle.dsl.*
 import org.jetbrains.kotlin.gradle.plugin.*
 import org.jetbrains.kotlin.gradle.plugin.mpp.KotlinNativeTarget
@@ -473,7 +473,13 @@ private class KlibValidationPipelineBuilder(
                 "into a single merged KLib ABI dump"
         dumpFileName = klibDumpFileName
         mergedFile = klibMergeDir.resolve(klibDumpFileName)
-        compilableTargets = project.compilableTargets()
+        val compilableTargets = project.compilableTargets()
+        filterTargets(Spec<String> {
+                name -> compilableTargets.get()
+            .map(KlibTarget::parse)
+            .map { it.configurableName }
+            .contains(name)
+        })
         onlyIf {
             compilableTargets.get().isNotEmpty()
         }
@@ -488,7 +494,13 @@ private class KlibValidationPipelineBuilder(
                 "different targets into a single merged KLib ABI dump"
         dumpFileName = klibDumpFileName
         mergedFile = klibMergeDir.resolve(klibDumpFileName)
-        compilableTargets = project.compilableTargets()
+        val compilableTargets = project.compilableTargets()
+        filterTargets(Spec<String> {
+            name -> compilableTargets.get()
+                .map(KlibTarget::parse)
+                .map { it.configurableName }
+                .contains(name)
+        })
         onlyIf {
             compilableTargets.get().isNotEmpty()
         }

--- a/src/main/kotlin/BinaryCompatibilityValidatorPlugin.kt
+++ b/src/main/kotlin/BinaryCompatibilityValidatorPlugin.kt
@@ -473,9 +473,6 @@ private class KlibValidationPipelineBuilder(
                 "into a single merged KLib ABI dump"
         dumpFileName = klibDumpFileName
         mergedFile = klibMergeDir.resolve(klibDumpFileName)
-        // At task configuration time, we don't know if a target will produce any artifacts,
-        // so we initialize it with all possible inputs and then filter them out during merge.
-        filterTargets(project.isTargetWithConfigurableNameCompilable())
         val compilableTargets = project.compilableTargets()
         onlyIf("There are no dumps to merge") {
             compilableTargets.get().isNotEmpty()
@@ -491,9 +488,6 @@ private class KlibValidationPipelineBuilder(
                 "different targets into a single merged KLib ABI dump"
         dumpFileName = klibDumpFileName
         mergedFile = klibMergeDir.resolve(klibDumpFileName)
-        // At task configuration time, we don't know if a target will produce any artifacts,
-        // so we initialize it with all possible inputs and then filter them out during merge.
-        filterTargets(project.isTargetWithConfigurableNameCompilable())
         val compilableTargets = project.compilableTargets()
         onlyIf("There are no dumps to merge") {
             compilableTargets.get().isNotEmpty()

--- a/src/main/kotlin/BinaryCompatibilityValidatorPlugin.kt
+++ b/src/main/kotlin/BinaryCompatibilityValidatorPlugin.kt
@@ -219,7 +219,9 @@ private fun Project.configureKotlinCompilation(
 
     val apiBuild = task<KotlinApiBuildTask>(targetConfig.apiTaskName("Build")) {
         // Do not enable task for empty umbrella modules
-        isEnabled = apiCheckEnabled(projectName, extension) && compilation.hasAnySources()
+        isEnabled = apiCheckEnabled(projectName, extension)
+        val hasSourcesPredicate = compilation.hasAnySourcesPredicate()
+        onlyIf { hasSourcesPredicate.get() }
         // 'group' is not specified deliberately, so it will be hidden from ./gradlew tasks
         description =
             "Builds Kotlin API for 'main' compilations of $projectName. Complementary task and shouldn't be called manually"
@@ -601,7 +603,9 @@ private class KlibValidationPipelineBuilder(
         val buildTask = project.task<KotlinKlibAbiBuildTask>(targetConfig.apiTaskName("Build")) {
             target = targetConfig.targetName!!
             // Do not enable task for empty umbrella modules
-            isEnabled = klibAbiCheckEnabled(projectName, extension) && compilation.hasAnySources()
+            isEnabled = klibAbiCheckEnabled(projectName, extension)
+            val hasSourcesPredicate = compilation.hasAnySourcesPredicate()
+            onlyIf { hasSourcesPredicate.get() }
             // 'group' is not specified deliberately, so it will be hidden from ./gradlew tasks
             description = "Builds Kotlin KLib ABI dump for 'main' compilations of $projectName. " +
                     "Complementary task and shouldn't be called manually"
@@ -636,7 +640,9 @@ private class KlibValidationPipelineBuilder(
     ): TaskProvider<KotlinKlibInferAbiForUnsupportedTargetTask> {
         val targetName = targetConfig.targetName!!
         return project.task<KotlinKlibInferAbiForUnsupportedTargetTask>(targetConfig.apiTaskName("Infer")) {
-            isEnabled = klibAbiCheckEnabled(project.name, extension) && compilation.hasAnySources()
+            isEnabled = klibAbiCheckEnabled(project.name, extension)
+            val hasSourcesPredicate = compilation.hasAnySourcesPredicate()
+            onlyIf { hasSourcesPredicate.get() }
             description = "Try to infer the dump for unsupported target $targetName using dumps " +
                     "generated for supported targets."
             group = "other"
@@ -694,4 +700,8 @@ private val Project.klibDumpFileName: String
 
 private fun KotlinCompilation<KotlinCommonOptions>.hasAnySources(): Boolean = allKotlinSourceSets.any {
     it.kotlin.srcDirs.any(File::exists)
+}
+
+private fun KotlinCompilation<KotlinCommonOptions>.hasAnySourcesPredicate(): Provider<Boolean> = project.provider {
+    this.hasAnySources()
 }

--- a/src/main/kotlin/BinaryCompatibilityValidatorPlugin.kt
+++ b/src/main/kotlin/BinaryCompatibilityValidatorPlugin.kt
@@ -417,7 +417,7 @@ private class KlibValidationPipelineBuilder(
                 project.name
         projectApiFile = klibApiDir.get().resolve(klibDumpFileName)
         generatedApiFile = klibMergeDir.resolve(klibDumpFileName)
-        val hasCompilableTargets = project.compilableTargetsPredicate()
+        val hasCompilableTargets = project.hasCompilableTargetsPredicate()
         onlyIf("There are no klibs compiled for the project") { hasCompilableTargets.get() }
     }
 
@@ -431,7 +431,7 @@ private class KlibValidationPipelineBuilder(
         group = "other"
         from = klibMergeDir.resolve(klibDumpFileName)
         to = klibApiDir.get().resolve(klibDumpFileName)
-        val hasCompilableTargets = project.compilableTargetsPredicate()
+        val hasCompilableTargets = project.hasCompilableTargetsPredicate()
         onlyIf("There are no klibs compiled for the project") { hasCompilableTargets.get() }
     }
 
@@ -451,7 +451,7 @@ private class KlibValidationPipelineBuilder(
         supportedTargets = supportedTargets()
         inputAbiFile = klibApiDir.get().resolve(klibDumpFileName)
         outputAbiFile = klibOutputDir.resolve(klibDumpFileName)
-        val hasCompilableTargets = project.compilableTargetsPredicate()
+        val hasCompilableTargets = project.hasCompilableTargetsPredicate()
         onlyIf("There are no klibs compiled for the project") { hasCompilableTargets.get() }
     }
 
@@ -468,7 +468,7 @@ private class KlibValidationPipelineBuilder(
                 "into a single merged KLib ABI dump"
         dumpFileName = klibDumpFileName
         mergedFile = klibMergeDir.resolve(klibDumpFileName)
-        val hasCompilableTargets = project.compilableTargetsPredicate()
+        val hasCompilableTargets = project.hasCompilableTargetsPredicate()
         onlyIf("There are no dumps to merge") { hasCompilableTargets.get() }
     }
 
@@ -481,7 +481,7 @@ private class KlibValidationPipelineBuilder(
                 "different targets into a single merged KLib ABI dump"
         dumpFileName = klibDumpFileName
         mergedFile = klibMergeDir.resolve(klibDumpFileName)
-        val hasCompilableTargets = project.compilableTargetsPredicate()
+        val hasCompilableTargets = project.hasCompilableTargetsPredicate()
         onlyIf("There are no dumps to merge") { hasCompilableTargets.get() }
     }
 
@@ -585,7 +585,7 @@ private class KlibValidationPipelineBuilder(
     }
 
     // Returns a predicate that checks if there are any compilable targets
-    private fun Project.compilableTargetsPredicate(): Provider<Boolean> {
+    private fun Project.hasCompilableTargetsPredicate(): Provider<Boolean> {
         return project.provider {
             project.kotlinMultiplatform.targets.matching { it.emitsKlib }
                 .asSequence()

--- a/src/main/kotlin/BinaryCompatibilityValidatorPlugin.kt
+++ b/src/main/kotlin/BinaryCompatibilityValidatorPlugin.kt
@@ -11,6 +11,7 @@ import org.gradle.api.*
 import org.gradle.api.plugins.*
 import org.gradle.api.provider.*
 import org.gradle.api.tasks.*
+import org.jetbrains.kotlin.cli.jvm.main
 import org.jetbrains.kotlin.gradle.dsl.*
 import org.jetbrains.kotlin.gradle.plugin.*
 import org.jetbrains.kotlin.gradle.plugin.mpp.KotlinNativeTarget
@@ -105,7 +106,7 @@ public class BinaryCompatibilityValidatorPlugin : Plugin<Project> {
         kotlin.targets.matching { it.jvmBased }.all { target ->
             val targetConfig = TargetConfig(project, extension, target.name, jvmDirConfig)
             if (target.platformType == KotlinPlatformType.jvm) {
-                target.mainCompilations.all {
+                target.mainCompilationOrNull?.also {
                     project.configureKotlinCompilation(it, extension, targetConfig, commonApiDump, commonApiCheck)
                 }
             } else if (target.platformType == KotlinPlatformType.androidJvm) {
@@ -219,11 +220,7 @@ private fun Project.configureKotlinCompilation(
 
     val apiBuild = task<KotlinApiBuildTask>(targetConfig.apiTaskName("Build")) {
         // Do not enable task for empty umbrella modules
-        isEnabled =
-            apiCheckEnabled(
-                projectName,
-                extension
-            ) && compilation.allKotlinSourceSets.any { it.kotlin.srcDirs.any { it.exists() } }
+        isEnabled = apiCheckEnabled(projectName, extension) && compilation.hasAnySources()
         // 'group' is not specified deliberately, so it will be hidden from ./gradlew tasks
         description =
             "Builds Kotlin API for 'main' compilations of $projectName. Complementary task and shouldn't be called manually"
@@ -419,6 +416,10 @@ private class KlibValidationPipelineBuilder(
                 project.name
         projectApiFile = klibApiDir.get().resolve(klibDumpFileName)
         generatedApiFile = klibMergeDir.resolve(klibDumpFileName)
+        val compilableTargets = project.compilableTargets()
+        onlyIf {
+            compilableTargets.get().isNotEmpty()
+        }
     }
 
     private fun Project.dumpKlibsTask(
@@ -431,6 +432,10 @@ private class KlibValidationPipelineBuilder(
         group = "other"
         from = klibMergeDir.resolve(klibDumpFileName)
         to = klibApiDir.get().resolve(klibDumpFileName)
+        val compTargets = project.compilableTargets()
+        onlyIf {
+            compTargets.get().isNotEmpty()
+        }
     }
 
     private fun Project.extractAbi(
@@ -449,6 +454,10 @@ private class KlibValidationPipelineBuilder(
         supportedTargets = supportedTargets()
         inputAbiFile = klibApiDir.get().resolve(klibDumpFileName)
         outputAbiFile = klibOutputDir.resolve(klibDumpFileName)
+        val compilableTargets = project.compilableTargets()
+        onlyIf {
+            compilableTargets.get().isNotEmpty()
+        }
     }
 
     private fun Project.mergeInferredKlibsUmbrellaTask(
@@ -464,6 +473,10 @@ private class KlibValidationPipelineBuilder(
                 "into a single merged KLib ABI dump"
         dumpFileName = klibDumpFileName
         mergedFile = klibMergeDir.resolve(klibDumpFileName)
+        compilableTargets = project.compilableTargets()
+        onlyIf {
+            compilableTargets.get().isNotEmpty()
+        }
     }
 
     private fun Project.mergeKlibsUmbrellaTask(
@@ -475,6 +488,10 @@ private class KlibValidationPipelineBuilder(
                 "different targets into a single merged KLib ABI dump"
         dumpFileName = klibDumpFileName
         mergedFile = klibMergeDir.resolve(klibDumpFileName)
+        compilableTargets = project.compilableTargets()
+        onlyIf {
+            compilableTargets.get().isNotEmpty()
+        }
     }
 
     fun Project.bannedTargets(): Set<String> {
@@ -499,10 +516,7 @@ private class KlibValidationPipelineBuilder(
 
         val supportedTargetsProvider = supportedTargets()
         kotlin.targets.matching { it.emitsKlib }.configureEach { currentTarget ->
-            val mainCompilations = currentTarget.mainCompilations
-            if (mainCompilations.none()) {
-                return@configureEach
-            }
+            val mainCompilation = currentTarget.mainCompilationOrNull ?: return@configureEach
 
             val targetName = currentTarget.targetName
             val targetConfig = TargetConfig(project, extension, targetName, intermediateFilesConfig)
@@ -510,19 +524,14 @@ private class KlibValidationPipelineBuilder(
             val targetSupported = targetIsSupported(currentTarget)
             // If a target is supported, the workflow is simple: create a dump, then merge it along with other dumps.
             if (targetSupported) {
-                mainCompilations.all {
-                    val buildTargetAbi = configureKlibCompilation(
-                        it, extension, targetConfig,
-                        apiBuildDir
-                    )
-                    mergeTask.configure {
-                        it.addInput(targetName, apiBuildDir)
-                        it.dependsOn(buildTargetAbi)
-                    }
-                    mergeInferredTask.configure {
-                        it.addInput(targetName, apiBuildDir)
-                        it.dependsOn(buildTargetAbi)
-                    }
+                val buildTargetAbi = configureKlibCompilation(mainCompilation, extension, targetConfig, apiBuildDir)
+                mergeTask.configure {
+                    it.addInput(targetName, apiBuildDir)
+                    it.dependsOn(buildTargetAbi)
+                }
+                mergeInferredTask.configure {
+                    it.addInput(targetName, apiBuildDir)
+                    it.dependsOn(buildTargetAbi)
                 }
                 return@configureEach
             }
@@ -534,9 +543,12 @@ private class KlibValidationPipelineBuilder(
             }
             // The actual merge will happen here, where we'll try to infer a dump for the unsupported target and merge
             // it with other supported target dumps.
-            val proxy = unsupportedTargetDumpProxy(klibApiDir, targetConfig,
+            val proxy = unsupportedTargetDumpProxy(
+                mainCompilation,
+                klibApiDir, targetConfig,
                 extractUnderlyingTarget(currentTarget),
-                apiBuildDir, supportedTargetsProvider)
+                apiBuildDir, supportedTargetsProvider
+            )
             mergeInferredTask.configure {
                 it.addInput(targetName, apiBuildDir)
                 it.dependsOn(proxy)
@@ -555,18 +567,20 @@ private class KlibValidationPipelineBuilder(
 
     private fun Project.targetIsSupported(target: KotlinTarget): Boolean {
         if (bannedTargets().contains(target.targetName)) return false
-        return when(target) {
+        return when (target) {
             is KotlinNativeTarget -> HostManager().isEnabled(target.konanTarget)
             else -> true
         }
     }
 
+    // Compilable targets supported by the host compiler
     private fun Project.supportedTargets(): Provider<Set<String>> {
         val banned = bannedTargets() // for testing only
         return project.provider {
             val hm = HostManager()
             project.kotlinMultiplatform.targets.matching { it.emitsKlib }
                 .asSequence()
+                .filter { it.mainCompilationOrNull?.hasAnySources() == true }
                 .filter {
                     if (it is KotlinNativeTarget) {
                         hm.isEnabled(it.konanTarget) && it.targetName !in banned
@@ -579,6 +593,16 @@ private class KlibValidationPipelineBuilder(
         }
     }
 
+    // Targets having some sources to compile
+    private fun Project.compilableTargets(): Provider<Set<String>> {
+        return project.provider {
+            project.kotlinMultiplatform.targets.matching { it.emitsKlib }
+                .asSequence()
+                .filter { it.mainCompilationOrNull?.hasAnySources() == true }
+                .map { KlibTarget(extractUnderlyingTarget(it), it.targetName).toString() }
+                .toSet()
+        }
+    }
 
     private fun Project.configureKlibCompilation(
         compilation: KotlinCompilation<KotlinCommonOptions>,
@@ -590,11 +614,7 @@ private class KlibValidationPipelineBuilder(
         val buildTask = project.task<KotlinKlibAbiBuildTask>(targetConfig.apiTaskName("Build")) {
             target = targetConfig.targetName!!
             // Do not enable task for empty umbrella modules
-            isEnabled =
-                klibAbiCheckEnabled(
-                    projectName,
-                    extension
-                ) && compilation.allKotlinSourceSets.any { it.kotlin.srcDirs.any { it.exists() } }
+            isEnabled = klibAbiCheckEnabled(projectName, extension) && compilation.hasAnySources()
             // 'group' is not specified deliberately, so it will be hidden from ./gradlew tasks
             description = "Builds Kotlin KLib ABI dump for 'main' compilations of $projectName. " +
                     "Complementary task and shouldn't be called manually"
@@ -620,6 +640,7 @@ private class KlibValidationPipelineBuilder(
     }
 
     private fun Project.unsupportedTargetDumpProxy(
+        compilation: KotlinCompilation<KotlinCommonOptions>,
         klibApiDir: Provider<File>,
         targetConfig: TargetConfig,
         underlyingTarget: String,
@@ -628,7 +649,7 @@ private class KlibValidationPipelineBuilder(
     ): TaskProvider<KotlinKlibInferAbiForUnsupportedTargetTask> {
         val targetName = targetConfig.targetName!!
         return project.task<KotlinKlibInferAbiForUnsupportedTargetTask>(targetConfig.apiTaskName("Infer")) {
-            isEnabled = klibAbiCheckEnabled(project.name, extension)
+            isEnabled = klibAbiCheckEnabled(project.name, extension) && compilation.hasAnySources()
             description = "Try to infer the dump for unsupported target $targetName using dumps " +
                     "generated for supported targets."
             group = "other"
@@ -676,10 +697,14 @@ private fun extractUnderlyingTarget(target: KotlinTarget): String {
 private val Project.kotlinMultiplatform
     get() = extensions.getByName("kotlin") as KotlinMultiplatformExtension
 
-private val KotlinTarget.mainCompilations
-    get() = compilations.matching { it.name == "main" }
+private val KotlinTarget.mainCompilationOrNull: KotlinCompilation<KotlinCommonOptions>?
+    get() = compilations.firstOrNull { it.name == KotlinCompilation.MAIN_COMPILATION_NAME }
 
 private val Project.jvmDumpFileName: String
     get() = "$name.api"
 private val Project.klibDumpFileName: String
     get() = "$name.klib.api"
+
+private fun KotlinCompilation<KotlinCommonOptions>.hasAnySources(): Boolean = allKotlinSourceSets.any {
+    it.kotlin.srcDirs.any(File::exists)
+}

--- a/src/main/kotlin/BinaryCompatibilityValidatorPlugin.kt
+++ b/src/main/kotlin/BinaryCompatibilityValidatorPlugin.kt
@@ -10,7 +10,6 @@ import kotlinx.validation.api.klib.konanTargetNameMapping
 import org.gradle.api.*
 import org.gradle.api.plugins.*
 import org.gradle.api.provider.*
-import org.gradle.api.specs.Spec
 import org.gradle.api.tasks.*
 import org.jetbrains.kotlin.gradle.dsl.*
 import org.jetbrains.kotlin.gradle.plugin.*
@@ -416,10 +415,8 @@ private class KlibValidationPipelineBuilder(
                 project.name
         projectApiFile = klibApiDir.get().resolve(klibDumpFileName)
         generatedApiFile = klibMergeDir.resolve(klibDumpFileName)
-        val compilableTargets = project.compilableTargets()
-        onlyIf("There are no klibs compiled for the project") {
-            compilableTargets.get().isNotEmpty()
-        }
+        val hasCompilableTargets = project.compilableTargetsPredicate()
+        onlyIf("There are no klibs compiled for the project") { hasCompilableTargets.get() }
     }
 
     private fun Project.dumpKlibsTask(
@@ -432,10 +429,8 @@ private class KlibValidationPipelineBuilder(
         group = "other"
         from = klibMergeDir.resolve(klibDumpFileName)
         to = klibApiDir.get().resolve(klibDumpFileName)
-        val compTargets = project.compilableTargets()
-        onlyIf("There are no klibs compiled for the project") {
-            compTargets.get().isNotEmpty()
-        }
+        val hasCompilableTargets = project.compilableTargetsPredicate()
+        onlyIf("There are no klibs compiled for the project") { hasCompilableTargets.get() }
     }
 
     private fun Project.extractAbi(
@@ -454,10 +449,8 @@ private class KlibValidationPipelineBuilder(
         supportedTargets = supportedTargets()
         inputAbiFile = klibApiDir.get().resolve(klibDumpFileName)
         outputAbiFile = klibOutputDir.resolve(klibDumpFileName)
-        val compilableTargets = project.compilableTargets()
-        onlyIf("There are no klibs compiled for the project") {
-            compilableTargets.get().isNotEmpty()
-        }
+        val hasCompilableTargets = project.compilableTargetsPredicate()
+        onlyIf("There are no klibs compiled for the project") { hasCompilableTargets.get() }
     }
 
     private fun Project.mergeInferredKlibsUmbrellaTask(
@@ -473,10 +466,8 @@ private class KlibValidationPipelineBuilder(
                 "into a single merged KLib ABI dump"
         dumpFileName = klibDumpFileName
         mergedFile = klibMergeDir.resolve(klibDumpFileName)
-        val compilableTargets = project.compilableTargets()
-        onlyIf("There are no dumps to merge") {
-            compilableTargets.get().isNotEmpty()
-        }
+        val hasCompilableTargets = project.compilableTargetsPredicate()
+        onlyIf("There are no dumps to merge") { hasCompilableTargets.get() }
     }
 
     private fun Project.mergeKlibsUmbrellaTask(
@@ -488,10 +479,8 @@ private class KlibValidationPipelineBuilder(
                 "different targets into a single merged KLib ABI dump"
         dumpFileName = klibDumpFileName
         mergedFile = klibMergeDir.resolve(klibDumpFileName)
-        val compilableTargets = project.compilableTargets()
-        onlyIf("There are no dumps to merge") {
-            compilableTargets.get().isNotEmpty()
-        }
+        val hasCompilableTargets = project.compilableTargetsPredicate()
+        onlyIf("There are no dumps to merge") { hasCompilableTargets.get() }
     }
 
     fun Project.bannedTargets(): Set<String> {
@@ -593,28 +582,12 @@ private class KlibValidationPipelineBuilder(
         }
     }
 
-    // Targets having some sources to compile
-    private fun Project.compilableTargets(): Provider<Set<String>> {
+    // Returns a predicate that checks if there are any compilable targets
+    private fun Project.compilableTargetsPredicate(): Provider<Boolean> {
         return project.provider {
             project.kotlinMultiplatform.targets.matching { it.emitsKlib }
                 .asSequence()
-                .filter { it.mainCompilationOrNull?.hasAnySources() == true }
-                .map { KlibTarget(extractUnderlyingTarget(it), it.targetName).toString() }
-                .toSet()
-        }
-    }
-
-    private fun Project.isTargetWithConfigurableNameCompilable(): Spec<String> {
-        val provider = compilableTargets()
-        return object : Spec<String> {
-            var configurableTargetNames: Set<String>? = null
-
-            override fun isSatisfiedBy(element: String?): Boolean {
-                if (configurableTargetNames == null) {
-                    configurableTargetNames = provider.get().map { KlibTarget.parse(it).configurableName }.toSet()
-                }
-                return configurableTargetNames!!.contains(element)
-            }
+                .any { it.mainCompilationOrNull?.hasAnySources() == true }
         }
     }
 

--- a/src/main/kotlin/BinaryCompatibilityValidatorPlugin.kt
+++ b/src/main/kotlin/BinaryCompatibilityValidatorPlugin.kt
@@ -417,7 +417,7 @@ private class KlibValidationPipelineBuilder(
         projectApiFile = klibApiDir.get().resolve(klibDumpFileName)
         generatedApiFile = klibMergeDir.resolve(klibDumpFileName)
         val compilableTargets = project.compilableTargets()
-        onlyIf {
+        onlyIf("There are no klibs compiled for the project") {
             compilableTargets.get().isNotEmpty()
         }
     }
@@ -433,7 +433,7 @@ private class KlibValidationPipelineBuilder(
         from = klibMergeDir.resolve(klibDumpFileName)
         to = klibApiDir.get().resolve(klibDumpFileName)
         val compTargets = project.compilableTargets()
-        onlyIf {
+        onlyIf("There are no klibs compiled for the project") {
             compTargets.get().isNotEmpty()
         }
     }
@@ -455,7 +455,7 @@ private class KlibValidationPipelineBuilder(
         inputAbiFile = klibApiDir.get().resolve(klibDumpFileName)
         outputAbiFile = klibOutputDir.resolve(klibDumpFileName)
         val compilableTargets = project.compilableTargets()
-        onlyIf {
+        onlyIf("There are no klibs compiled for the project") {
             compilableTargets.get().isNotEmpty()
         }
     }
@@ -473,14 +473,11 @@ private class KlibValidationPipelineBuilder(
                 "into a single merged KLib ABI dump"
         dumpFileName = klibDumpFileName
         mergedFile = klibMergeDir.resolve(klibDumpFileName)
+        // At task configuration time, we don't know if a target will produce any artifacts,
+        // so we initialize it with all possible inputs and then filter them out during merge.
+        filterTargets(project.isTargetWithConfigurableNameCompilable())
         val compilableTargets = project.compilableTargets()
-        filterTargets(Spec<String> {
-                name -> compilableTargets.get()
-            .map(KlibTarget::parse)
-            .map { it.configurableName }
-            .contains(name)
-        })
-        onlyIf {
+        onlyIf("There are no dumps to merge") {
             compilableTargets.get().isNotEmpty()
         }
     }
@@ -494,14 +491,11 @@ private class KlibValidationPipelineBuilder(
                 "different targets into a single merged KLib ABI dump"
         dumpFileName = klibDumpFileName
         mergedFile = klibMergeDir.resolve(klibDumpFileName)
+        // At task configuration time, we don't know if a target will produce any artifacts,
+        // so we initialize it with all possible inputs and then filter them out during merge.
+        filterTargets(project.isTargetWithConfigurableNameCompilable())
         val compilableTargets = project.compilableTargets()
-        filterTargets(Spec<String> {
-            name -> compilableTargets.get()
-                .map(KlibTarget::parse)
-                .map { it.configurableName }
-                .contains(name)
-        })
-        onlyIf {
+        onlyIf("There are no dumps to merge") {
             compilableTargets.get().isNotEmpty()
         }
     }
@@ -613,6 +607,20 @@ private class KlibValidationPipelineBuilder(
                 .filter { it.mainCompilationOrNull?.hasAnySources() == true }
                 .map { KlibTarget(extractUnderlyingTarget(it), it.targetName).toString() }
                 .toSet()
+        }
+    }
+
+    private fun Project.isTargetWithConfigurableNameCompilable(): Spec<String> {
+        val provider = compilableTargets()
+        return object : Spec<String> {
+            var configurableTargetNames: Set<String>? = null
+
+            override fun isSatisfiedBy(element: String?): Boolean {
+                if (configurableTargetNames == null) {
+                    configurableTargetNames = provider.get().map { KlibTarget.parse(it).configurableName }.toSet()
+                }
+                return configurableTargetNames!!.contains(element)
+            }
         }
     }
 

--- a/src/main/kotlin/KotlinKlibMergeAbiTask.kt
+++ b/src/main/kotlin/KotlinKlibMergeAbiTask.kt
@@ -6,8 +6,10 @@
 package kotlinx.validation
 
 import kotlinx.validation.api.klib.KlibDump
+import kotlinx.validation.api.klib.KlibTarget
 import kotlinx.validation.api.klib.saveTo
 import org.gradle.api.DefaultTask
+import org.gradle.api.provider.Provider
 import org.gradle.api.tasks.*
 import java.io.File
 
@@ -45,6 +47,12 @@ internal abstract class KotlinKlibMergeAbiTask : DefaultTask() {
     @Input
     lateinit var dumpFileName: String
 
+    /**
+     * Provides serialized [KlibTarget]s having some sources to compile.
+     */
+    @Input
+    lateinit var compilableTargets: Provider<Set<String>>
+
     internal fun addInput(target: String, file: File) {
         targetToFile[target] = file
     }
@@ -52,9 +60,15 @@ internal abstract class KotlinKlibMergeAbiTask : DefaultTask() {
     @OptIn(ExperimentalBCVApi::class)
     @TaskAction
     internal fun merge() {
+        val compilableTargetNames = compilableTargets.get().map {
+            KlibTarget.parse(it).configurableName
+        }.toSet()
+
         KlibDump().apply {
             targetToFile.forEach { (targetName, dumpDir) ->
-                merge(dumpDir.resolve(dumpFileName), targetName)
+                if (compilableTargetNames.contains(targetName)) {
+                    merge(dumpDir.resolve(dumpFileName), targetName)
+                }
             }
         }.saveTo(mergedFile)
     }

--- a/src/main/kotlin/KotlinKlibMergeAbiTask.kt
+++ b/src/main/kotlin/KotlinKlibMergeAbiTask.kt
@@ -8,7 +8,6 @@ package kotlinx.validation
 import kotlinx.validation.api.klib.KlibDump
 import kotlinx.validation.api.klib.saveTo
 import org.gradle.api.DefaultTask
-import org.gradle.api.specs.Spec
 import org.gradle.api.tasks.*
 import java.io.File
 
@@ -46,24 +45,18 @@ internal abstract class KotlinKlibMergeAbiTask : DefaultTask() {
     @Input
     lateinit var dumpFileName: String
 
-    private var targetPredicate: Spec<String> = Spec<String> { true }
-
     internal fun addInput(target: String, file: File) {
         targetToFile[target] = file
-    }
-
-    internal fun filterTargets(filter: Spec<String>) {
-        targetPredicate = filter
     }
 
     @OptIn(ExperimentalBCVApi::class)
     @TaskAction
     internal fun merge() {
-        val filter = targetPredicate
         KlibDump().apply {
             targetToFile.forEach { (targetName, dumpDir) ->
-                if (filter.isSatisfiedBy(targetName)) {
-                    merge(dumpDir.resolve(dumpFileName), targetName)
+                val dumpFile = dumpDir.resolve(dumpFileName)
+                if (dumpFile.exists()) {
+                    merge(dumpFile, targetName)
                 }
             }
         }.saveTo(mergedFile)

--- a/src/main/kotlin/KotlinKlibMergeAbiTask.kt
+++ b/src/main/kotlin/KotlinKlibMergeAbiTask.kt
@@ -6,10 +6,9 @@
 package kotlinx.validation
 
 import kotlinx.validation.api.klib.KlibDump
-import kotlinx.validation.api.klib.KlibTarget
 import kotlinx.validation.api.klib.saveTo
 import org.gradle.api.DefaultTask
-import org.gradle.api.provider.Provider
+import org.gradle.api.specs.Spec
 import org.gradle.api.tasks.*
 import java.io.File
 
@@ -47,26 +46,23 @@ internal abstract class KotlinKlibMergeAbiTask : DefaultTask() {
     @Input
     lateinit var dumpFileName: String
 
-    /**
-     * Provides serialized [KlibTarget]s having some sources to compile.
-     */
-    @Input
-    lateinit var compilableTargets: Provider<Set<String>>
+    private var targetPredicate: Spec<String> = Spec<String> { true }
 
     internal fun addInput(target: String, file: File) {
         targetToFile[target] = file
     }
 
+    internal fun filterTargets(filter: Spec<String>) {
+        targetPredicate = filter
+    }
+
     @OptIn(ExperimentalBCVApi::class)
     @TaskAction
     internal fun merge() {
-        val compilableTargetNames = compilableTargets.get().map {
-            KlibTarget.parse(it).configurableName
-        }.toSet()
-
+        val filter = targetPredicate
         KlibDump().apply {
             targetToFile.forEach { (targetName, dumpDir) ->
-                if (compilableTargetNames.contains(targetName)) {
+                if (filter.isSatisfiedBy(targetName)) {
                     merge(dumpDir.resolve(dumpFileName), targetName)
                 }
             }

